src/byte_array.rs

```diff
@@ -0,0 +1,264 @@
+                bytes.as_ptr() as *const _,
+                bytes.len() as u32,
+            );
+        }
+        self
+    }
+
+    pub fn prepend<T: AsRef<[u8]>>(&self, data: T) -> &Self {
+        let bytes = data.as_ref();
+        unsafe {
+            glib_sys::g_byte_array_prepend(
+                self.to_glib_none().0,
+                bytes.as_ptr() as *const _,
+                bytes.len() as u32,
+            );
+        }
+        self
+    }
+
+    pub fn remove_index(&self, index: usize) {
+        unsafe {
+            glib_sys::g_byte_array_remove_index(self.to_glib_none().0, index as u32);
+        }
+    }
+
+    pub fn remove_index_fast(&self, index: usize) {
+        unsafe {
+            glib_sys::g_byte_array_remove_index_fast(self.to_glib_none().0, index as u32);
+        }
+    }
+
+    pub fn remove_range(&self, index: usize, length: usize) {
+        unsafe {
+            glib_sys::g_byte_array_remove_range(self.to_glib_none().0, index as u32, length as u32);
+        }
+    }
+
+    pub fn set_size(&self, size: usize) {
+        unsafe {
+            glib_sys::g_byte_array_set_size(self.to_glib_none().0, size as u32);
+        }
+    }
+
+    pub fn sort<F: FnMut(&u8, &u8) -> Ordering>(&self, compare_func: F) {
+        unsafe {
+            let mut func = compare_func;
+            let func_obj: &mut (FnMut(&u8, &u8) -> Ordering) = &mut func;
+            let func_ptr =
+                &func_obj as *const &mut (FnMut(&u8, &u8) -> Ordering) as glib_sys::gpointer;
+
+            glib_sys::g_byte_array_sort_with_data(
+                self.to_glib_none().0,
+                Some(compare_func_trampoline),
+                func_ptr,
+            );
+        }
+    }
+}
+
+unsafe extern "C" fn compare_func_trampoline(
+    a: glib_sys::gconstpointer,
+    b: glib_sys::gconstpointer,
+    func: glib_sys::gpointer,
+) -> i32 {
+    let func = func as *mut &mut (FnMut(&u8, &u8) -> Ordering);
+
+    let a = &*(a as *const u8);
+    let b = &*(b as *const u8);
+
+    match (*func)(&a, &b) {
+        Ordering::Less => -1,
+        Ordering::Equal => 0,
+        Ordering::Greater => 1,
+    }
+}
+
+impl AsRef<[u8]> for ByteArray {
+    fn as_ref(&self) -> &[u8] {
+        &*self
+    }
+}
+
+impl<'a, T: ?Sized + Borrow<[u8]> + 'a> From<&'a T> for ByteArray {
+    fn from(value: &'a T) -> ByteArray {
+        let ba = ByteArray::new();
+        ba.append(value.borrow());
+        ba
+    }
+}
+
+impl Deref for ByteArray {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        unsafe {
+            let ptr = (*self.to_glib_none().0).data;
+            let len = (*self.to_glib_none().0).len as usize;
+            debug_assert!(!ptr.is_null() || len == 0);
+            slice::from_raw_parts(ptr as *const u8, len)
+        }
+    }
+}
+
+impl Default for ByteArray {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+unsafe impl Send for ByteArray {}
+unsafe impl Sync for ByteArray {}
```

Comment:
`ByteArray` is a thread-safe type?

Reply:
as much as GString, which also impl Send/Sync.

but only refcounting seems safe

Reply:
It would only be Send/Sync if it used normal Rust mutability rules and no reference counting

Reply:
I should drop it then?

Reply:
Yes, it's wrong currently

Reply:
ok, removed